src/Utilities/FileSystem.hpp

```diff
@@ -0,0 +1,125 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Declares functions to do file system manipulations
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+/*!
+ * \ingroup FileSystem
+ * \brief A light-weight file system library based on POSIX.
+ *
+ * We use this library instead of a subprocess based library because OpenMPI
+ * does not support forking of processes on all systems. Since the
+ * parallelization library we use may be implemented on top of OpenMPI we
+ * take the safe route and use POSIX.
+ */
+namespace filesystem {
+
+/*!
+ * \ingroup FileSystem
+ * \brief Given a path possibly including a file name returns the path
```

Comment:
The Python `os.path` module has various routines which may be interesting, if only as an example of functionality that someone considered useful: https://docs.python.org/2/library/os.path.html.

In particular, I think `os.path.split` may have a similar functionality to what you were trying to implement here. Combining the two functions into one reduces some of the ambiguity, but does not totally remove it.

I suspect there is some version of this function that is generically useful enough to be worth implementing (vs. every user writing their own `string.find()` manipulations explicitly), but I am not sure which version that would be.